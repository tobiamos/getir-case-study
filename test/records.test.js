const request = require('supertest');
const mongoose = require('mongoose');
const { wipeMongo } = require('./utils');
const models = require('../lib/models');
const app = require('../app');

describe('records api', () => {
  beforeEach(async () => {
    await wipeMongo();
  });

  test('filter with empty request body', async () => {
    const res = await request(app)
      .post('/api/v1/records')
      .send({})
      .expect(400);

    expect(res.body).toMatchSnapshot();
  });

  test('filter with wrong date format', async () => {
    const res = await request(app)
      .post('/api/v1/records')
      .send({
        startDate: '2020/02/01',
        endDate: '2020-02-01',
        minCount: 200,
        maxCount: 500,
      });

    expect(res.body).toMatchSnapshot();
  });

  test('filter with end date before start date', async () => {
    const res = await request(app)
      .post('/api/v1/records')
      .send({
        startDate: '2020-02-02',
        endDate: '2020-02-01',
        minCount: 200,
        maxCount: 500,
      });

    expect(res.body).toMatchSnapshot();
  });

  test('filter with min count less than 0', async () => {
    const res = await request(app)
      .post('/api/v1/records')
      .send({
        startDate: '2020-02-02',
        endDate: '2020-02-01',
        minCount: -1,
        maxCount: 500,
      });

    expect(res.body).toMatchSnapshot();
  });

  test('filter with emtpy results in the response', async () => {
    const res = await request(app)
      .post('/api/v1/records')
      .send({
        startDate: '2020-02-01',
        endDate: '2020-02-01',
        minCount: 100,
        maxCount: 500,
      });

    expect(res.body).toMatchSnapshot();
  });

  test('filter with min count greater than max count', async () => {
    const res = await request(app)
      .post('/api/v1/records')
      .send({
        startDate: '2020-02-01',
        endDate: '2020-02-01',
        minCount: 1000,
        maxCount: 500,
      });

    expect(res.body).toMatchSnapshot();
  });

  test('filter with results in the response', async () => {
    await models.connect();

    const records = [{
      counts: [20, 30],
      key: 'VtXguuOz',
      createdAt: '2016-01-01T23:08:54.343Z',
    },
    {
      counts: [30, 3600],
      key: 'HjKCYMrZ',
      createdAt: '2016-01-01T18:57:47.340Z',
    },
    ];

    await mongoose.model('record').insertMany(records);

    const res = await request(app)
      .post('/api/v1/records')
      .send({
        startDate: '2012-01-01',
        endDate: '2017-02-01',
        minCount: 10,
        maxCount: 10000,
      });

    expect(res.body.records[0]).toMatchSnapshot({
      _id: expect.anything(),
    });
  });
});
