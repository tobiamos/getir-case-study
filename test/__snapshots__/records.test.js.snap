// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`records api filter with empty request body 1`] = `
Object {
  "code": 400,
  "msg": "Start Date is a required field",
  "records": Array [],
}
`;

exports[`records api filter with emtpy results in the response 1`] = `
Object {
  "code": 404,
  "msg": "Your query did not return any results :(",
  "records": Array [],
}
`;

exports[`records api filter with end date before start date 1`] = `
Object {
  "code": 400,
  "msg": "End Date must be after Start Date",
  "records": Array [],
}
`;

exports[`records api filter with min count greater than max count 1`] = `
Object {
  "code": 400,
  "msg": "Min count must be lesser than or equal to Max count",
  "records": Array [],
}
`;

exports[`records api filter with min count less than 0 1`] = `
Object {
  "code": 400,
  "msg": "Min count should have a minimum value of 0",
  "records": Array [],
}
`;

exports[`records api filter with results in the response 1`] = `
Object {
  "_id": Anything,
  "createdAt": "2016-01-01T23:08:54.343Z",
  "key": "VtXguuOz",
  "totalCount": 50,
}
`;

exports[`records api filter with wrong date format 1`] = `
Object {
  "code": 400,
  "msg": "Date must be in YYY-MM-DD format",
  "records": Array [],
}
`;
